### PR TITLE
Add parallel lattice rescoring using PyTorch-trained neural LMs

### PR DESCRIPTION
**Pull request by keli78**
Also, add examples and rescoring results with Transformer LMs for SWBD and WSJ.

### CODE REVIEW

File: egs/swbd/s5c/local/pytorchnn/data_prep.sh [hunk 1]
```diff
@@ -33,15 +33,16 @@ if [ $# != 1 ]; then
...
 
 for f in $train $valid $test $fisher; do
     [ ! -f $f ] && echo "$0: expected file $f to exist." && exit 1
 done
 
 # Sort and preprocess SWBD dataset
-local/pytorchnn/sort_by_start_time.py --infile $train --outfile $dir/swbd.train.sorted.txt
-local/pytorchnn/sort_by_start_time.py --infile $valid --outfile $dir/swbd.valid.sorted.txt
-local/pytorchnn/sort_by_start_time.py --infile $test --outfile $dir/swbd.test.sorted.txt
+python local/pytorchnn/sort_by_start_time.py --infile $train --outfile $dir/swbd.train.sorted.txt
```
Comment (danpovey) [inline egs/swbd/s5c/local/pytorchnn/data_prep.sh:43]: I prefer to always explicitly invoke python3, in case python2 is the default python.

File: src/latbin/lattice-expand.cc [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+#include "lat/lattice-functions.h"
+
+int main(int argc, char *argv[]) {
+  try {
+    using namespace kaldi;
+    typedef kaldi::int32 int32;
+    typedef kaldi::int64 int64;
+
+    const char *usage = 
+      "Expand lattices so that arcs with higher posteriors than expsilon,\n"
```
Comment (danpovey) [inline src/latbin/lattice-expand.cc:34]: expsilon misspelling.
I'd remove "which is a specific probability".

File: egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    $cmd JOB=1:$nj $dir/log/nnlmrescore.JOB.log \
+      lattice-add-nnlmscore --lm-scale=$nnlm_weight "ark:gunzip -c $expansion_dir/lat.JOB.gz|" \
+      $scores/arc_scores.JOB ark:- \| \
+      lattice-lmrescore --lm-scale=$oldlm_weight \
+      ark:- "$oldlm_command" "ark:|gzip -c>$dir/lat.JOB.gz" || exit 1;
+  else
+    $cmd JOB=1:$nj $dir/log/nnlmrescore.JOB.log \
+      lattice-add-nnlmscore --lm-scale=$nnlm_weight "ark:gunzip -c $expansion_dir/lat.JOB.gz|" \
+      $scores/arc_scores.JOB ark:- \| \
+      lattice-lmrescore-const-arpa --lm-scale=$oldlm_weight \
```
Comment (danpovey) [inline egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh:142]: What is the benefit of doing this versus just taking the best of the n-best-list elements?

File: egs/swbd/s5c/local/pytorchnn/data_prep.sh (hunk 1 above)
Comment (keli78) [reply egs/swbd/s5c/local/pytorchnn/data_prep.sh:43]: Sure. I will modify all related parts accordingly.

File: src/latbin/lattice-expand.cc (hunk 2 above)
Comment (keli78) [reply src/latbin/lattice-expand.cc:34]: ok, thanks!

File: egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh (hunk 3 above)
Comment (keli78) [reply egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh:142]: I did this instead of only taking the neural LM scores since usually interpolating the neural LM with the original n-gram model gives better WERs.

Comment (danpovey) [reply egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh:142]: .. I mean what is the observed difference in WER?

Comment (keli78) [reply egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh:142]: This gives an absolute 0.2% WER improvement on SWBD compared to w/o it.

Comment (keli78) [reply egs/wsj/s5/steps/pytorchnn/lmrescore_lattice_pytorchnn.sh:142]: ...and no difference on WSJ's eval92 test set. WER on eval92 after lattice rescoring is already very low (1.58% with a Transformer LM). That's probably why interpolating with the n-gram LM doesn't further help.

**Comment by keli78** [comment]
@danpovey I resolved the issues you mentioned. Let me know if you have more comments.

**Comment by danpovey** [comment]
Thanks!!  Merging.

**Comment by RuABraun** [comment]
This is nice. Would you mind sharing what sort of results you're getting?

**Comment by keli78** [comment]
Hi, @RuABraun, WERs on SWBD and WSJ can be found here: egs/swbd/s5c/local/pytorchnn/run_nnlm.sh and egs/wsj/s5/local/pytorchnn/run_nnlm.sh

**Comment by RuABraun** [comment]
Thank you! Any ideas why the results are not better than LSTMs? Lack of data?

**Comment by keli78** [comment]
@RuABraun I think the WERs from Transformer LMs are better than LSTMs on SWBD, though the perplexities are similar. Which dataset/results you're talking about?
